System data streams are not being upgraded in the feature migration API

alexey-ivanov-es opened the pull request:
This commit adds support for system data streams reindexing. The system data stream migration extends the existing system indices migration task and uses the [data stream reindex API](https://www.elastic.co/guide/en/elasticsearch/reference/8.x/data-stream-reindex-api.html).
The system index migration task starts a reindex data stream task and tracks its status every second. Only one system index or system data stream is migrated at a time. If a data stream migration fails, the entire system index migration task will also fail.
 
Closes: ES-10934
Closes: https://github.com/elastic/elasticsearch/issues/122949

alexey-ivanov-es committed:
System data streams are not being upgraded in the feature migration API

elasticsearchmachine commented:
Hi @alexey-ivanov-es, I've created a changelog YAML for you.

alexey-ivanov-es committed:
Update docs/changelog/123926.yaml

alexey-ivanov-es commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -134,7 +139,13 @@ public class SystemIndices {
         new Feature(TASKS_FEATURE_NAME, "Manages task results", List.of(TASKS_DESCRIPTOR)),
-        new Feature(SYNONYMS_FEATURE_NAME, "Manages synonyms", List.of(SYNONYMS_DESCRIPTOR))
+        new Feature(SYNONYMS_FEATURE_NAME, "Manages synonyms", List.of(SYNONYMS_DESCRIPTOR)),
+        new Feature("sds-1", "Manages system data stream", Collections.emptyList(),
```
oops, I'll remove this

alexey-ivanov-es committed:
Revert condition

Commit:
[CI] Auto commit changes from spotless

JVerwolf commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/shards/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -1211,29 +1211,6 @@ static List<Diagnosis.Resource> getRestoreFromSnapshotAffectedResources(
                 }
             }
 
-            Map<String, Set<String>> featureToDsBackingIndices = systemIndices.getFeatures()
-                .stream()
-                .collect(
-                    toMap(
-                        SystemIndices.Feature::getName,
-                        feature -> feature.getDataStreamDescriptors()
-                            .stream()
-                            .flatMap(descriptor -> descriptor.getBackingIndexNames(metadata).stream())
-                            .collect(toSet())
-                    )
-                );
-
-            // the shards_availability indicator works with indices so let's remove the feature states data streams backing indices from
-            // the list of affected indices (the feature state will cover the restore of these indices too)
-            for (Map.Entry<String, Set<String>> featureToBackingIndices : featureToDsBackingIndices.entrySet()) {
-                for (String featureIndex : featureToBackingIndices.getValue()) {
-                    if (restoreFromSnapshotIndices.contains(featureIndex)) {
-                        affectedFeatureStates.add(featureToBackingIndices.getKey());
-                        affectedIndices.remove(featureIndex);
-                    }
-                }
-            }
-
```
Why is this no longer needed?

alexey-ivanov-es replied:
At 1198 of this file `feature.getIndexDescriptors()` has changed to `feature.getSystemResourceDescriptors()` which includes both indices and data streams. The rest of the removed code was essentially the same as there, so now existing code covers both indices and data streams.

JVerwolf commented on server/src/main/java/org/elasticsearch/indices/IndexMatcher.java:
```diff
@@ -14,23 +14,15 @@
 /**
- * An IndexPatternMatcher holds an index pattern in a string and, given a
- * {@link Metadata} object, can return a list of index names matching that pattern.
+ * An IndexMatcher given a @link Metadata} object, can return a list of index names matching that pattern.
```
missing opening brace.

JVerwolf commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -134,7 +139,24 @@ public class SystemIndices {
     private static final Map<String, Feature> SERVER_SYSTEM_FEATURE_DESCRIPTORS = Stream.of(
         new Feature(TASKS_FEATURE_NAME, "Manages task results", List.of(TASKS_DESCRIPTOR)),
-        new Feature(SYNONYMS_FEATURE_NAME, "Manages synonyms", List.of(SYNONYMS_DESCRIPTOR))
+        new Feature(SYNONYMS_FEATURE_NAME, "Manages synonyms", List.of(SYNONYMS_DESCRIPTOR)),
+        new Feature(
+            "sds-1",
+            "Manages system data stream",
+            Collections.emptyList(),
+            Collections.singletonList(
+                new SystemDataStreamDescriptor(
+                    ".sds-1",
+                    "Test System Data Stream",
+                    SystemDataStreamDescriptor.Type.INTERNAL,
+                    ComposableIndexTemplate.builder().build(),
+                    Collections.emptyMap(),
+                    List.of("es"),
+                    "sds",
+                    ExecutorNames.DEFAULT_SYSTEM_DATA_STREAM_THREAD_POOLS
+                )
+            )
+        )
```
I'm guessing this is a leftover testing artifact?

alexey-ivanov-es replied:
Yep, I'll remove it

JVerwolf reviewed (commented):
I'm still mid-review, posting my interim comments.

alexey-ivanov-es committed:
Fix compilation and some review comments

JVerwolf commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -306,6 +339,10 @@ private void updateTaskState(SystemIndexMigrationInfo migrationInfo, Consumer<Cl
         }, this::markAsFailed));
     }
 
+    private static boolean needToBeMigrated(Stream<IndexMetadata> indicesMetadata) {
+        return indicesMetadata.anyMatch(SystemIndexMigrator::needsToBeMigrated);
+    }
+
```
Is there any need for this to be separate from `needsToBeMigrated`?  `needsToBeMigrated` is only used here.  WDYT about combining these?

alexey-ivanov-es replied:
Inlined

alexey-ivanov-es committed:
Fix compilation

alexey-ivanov-es committed:
Inline method

alexey-ivanov-es committed:
Fix compilation

alexey-ivanov-es committed:
Improve restarts handling

alexey-ivanov-es commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -575,12 +619,154 @@ private void reindex(SystemIndexMigrationInfo migrationInfo, ActionListener<Bulk
+                            // this might happen if the task has been migrated to another node
+                            // in this case we can just wait for the data stream migration task to finish
+                            logger.debug("data stream [{}] migration is already in progress", dataStreamName);
+                            checkDataStreamMigrationStatus(migrationInfo, completionListener);
```
Here we might need to cancel an existing task if it has failed...

Commit:
[CI] Auto commit changes from spotless

alexey-ivanov-es committed:
Improve restarts handling

alexey-ivanov-es committed:
fix error message

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

alexey-ivanov-es committed:
Improve comments

alexey-ivanov-es commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -575,12 +619,189 @@ private void reindex(SystemIndexMigrationInfo migrationInfo, ActionListener<Bulk
+                }
+            }, ex -> {
+                if (restartMigrationOnError) {
+                    cancelExistingDataStreamMigrationAndRetry(migrationInfo, completionListener);
```
https://www.elastic.co/guide/en/elasticsearch/reference/master/data-stream-reindex-api.html#reindex-data-stream-handling-failure

Commit:
[CI] Auto commit changes from spotless

JVerwolf commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -575,12 +619,188 @@ private void reindex(SystemIndexMigrationInfo migrationInfo, ActionListener<Bulk
+        SystemDataStreamMigrationInfo migrationInfo,
+        Consumer<SystemDataStreamMigrationInfo> completionListener
+    ) {
+        logger.debug(
```
Should this be at the warn level?

alexey-ivanov-es replied:
I don't think so. This code should only execute in one scenario: the migration of a data stream starts -> the [child task](https://github.com/elastic/elasticsearch/blob/8.x/x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/task/ReindexDataStreamPersistentTaskExecutor.java#L52) migrating the data stream fails for some reason -> the user restarts the system indices migration task.

Since the data stream migration task [retains](https://github.com/elastic/elasticsearch/blob/8.x/x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/task/ReindexDataStreamPersistentTaskExecutor.java#L295) its status for 1 day, we need to "cancel" the existing task and start a new one to avoid getting the same failure result again. In this case, the failure has already been logged.
"Canceling" the task here is part of normal operation, so I don't think it should be logged as error or warning.  I don't even want to make it info as it might be confusing to see

JVerwolf commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -575,12 +619,188 @@ private void reindex(SystemIndexMigrationInfo migrationInfo, ActionListener<Bulk
+
+        ActionListener<AcknowledgedResponse> listener = ActionListener.wrap(response -> {
+            if (response.isAcknowledged()) {
+                migrateDataStream(migrationInfo, completionListener);
```
Could this go into an infinite recursive loop if another part of the system is causing persistent errors? e.g. `migrateDataStream()` -> `checkDataStreamMigrationStatus()` -> `cancelExistingDataStreamMigrationAndRetry()` ->  `cancelDataStreamMigration()` -> `migrateDataStream()` -> ...
(I may be missunderstanding the flow control here)

In the original system index migration logic, an error would just stop the whole migration from continuing, such that a user would need to retry.  I spoke with @gwbrown, who mentioned that this was intentional in case there was a failure elsewhere in the system.  WDYT?

alexey-ivanov-es replied:
It shouldn't. The chain `migrateDataStream()` -> ... -> `cancelExistingDataStreamMigrationAndRetry()` is possible if a task for migrating that data stream already exists and is in a failed state. In that case, we cancel it and start a new one.
A new task is only started if the cancellation is successful. This means we shouldn't receive a `ResourceAlreadyExistsException` (line 649), which means that `checkDataStreamMigrationStatus` is called with `restartMigrationOnError=false` (line 647).

There is still slight possibility that we end up at the line 714 not because of exception from `dataStreamMigrationFailed` (lines 701 and 703), but because of another error, I will try to improve handling there.

alexey-ivanov-es replied:
>an error would just stop the whole migration from continuing

This logic is similar to deleting existing index created during failed migration attempt

JVerwolf replied:
> Logic here is similar to deleting existing index created during failed migration attempt

Yeah that makes sense and it's a good idea to retry.  

>There is still slight possibility that we end up at the line 714 not because of exception from dataStreamMigrationFailed (lines 701 and 703), but because of another error, I will try to improve handling there.

Exactly.  I'm worried that an issue in some other part of ES might cause this to just loop forever.

alexey-ivanov-es replied:
>Exactly. I'm worried that an issue in some other part of ES might cause this to just loop forever.

Should be fixed now

JVerwolf commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemResourceMigrationFactory.java:
```diff
@@ -0,0 +1,64 @@
+class SystemResourceMigrationFactory {
+    /**
+     * Convenience factory method holding the logic for creating instances from a Feature object.
+     * @param feature The feature that
```
> @param feature The feature that

The suspense is killing me!  😄

JVerwolf reviewed (commented):
Overall LGTM.  

I think this still needs some tests, but otherwise looks g2g. Have a look in FeatureMigrationIT for ideas.  Specifically, a test that fails the migration, similar to `testIndexBlockIsRemovedWhenAliasRequestFails` may be a good idea to ensure this doesn't get caught in a recursive loop (though ultimately your decision here).  

Nice work.

alexey-ivanov-es commented on server/src/main/java/org/elasticsearch/indices/SystemIndices.java:
```diff
@@ -882,6 +886,13 @@ public Collection<SystemDataStreamDescriptor> getDataStreamDescriptors() {
         }
 
+        /**
+         * Returns descriptors of all system resources - indices and data streams. Doesn't include associated indices.
```
No, it doesn't include backing indices. It refers to [org.elasticsearch.indices.AssociatedIndexDescriptor](https://github.com/elastic/elasticsearch/blob/8.x/server/src/main/java/org/elasticsearch/indices/AssociatedIndexDescriptor.java). I am not sure why we need them, but they exist. I'll update the comment to make it clearer

alexey-ivanov-es committed:
Address review comments

alexey-ivanov-es committed:
Fix security

JVerwolf commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
Yeah that's fair.

alexey-ivanov-es committed:
Improve handling of existing data stream migration task in failed state

alexey-ivanov-es committed:
Add some tests

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

Commit:
[CI] Auto commit changes from spotless

alexey-ivanov-es committed:
Add more tests

alexey-ivanov-es commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/action/TransportGetFeatureUpgradeStatusAction.java:
```diff
@@ -152,11 +152,31 @@ static List<GetFeatureUpgradeStatusResponse.IndexInfo> getIndexInfos(ClusterStat
+                Exception dsException = (descriptor.getDataStreamName().equals(failedResourceName)) ? exception : null;
+
+                // we don't know migration of which backing index has failed,
+                // so, unfortunately, have to report exception for all indices for now
```
Ugly solution, if you see more elegant one please don't hesitate to suggest

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

Commit:
[CI] Auto commit changes from spotless

alexey-ivanov-es committed:
Rename class

alexey-ivanov-es committed:
Fix error message if migration fails

Commit:
[CI] Auto commit changes from spotless

alexey-ivanov-es commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -575,12 +619,191 @@ private void reindex(SystemIndexMigrationInfo migrationInfo, ActionListener<Bulk
+        String dataStreamName = migrationInfo.getDataStreamName();
+        logger.info("migrating data stream [{}] from feature [{}]", dataStreamName, migrationInfo.getFeatureName());
+
+        ReindexDataStreamAction.ReindexDataStreamRequest reindexRequest = new ReindexDataStreamAction.ReindexDataStreamRequest(
```
> I don't think that system index migration is cancellable though, right? I assume that's by design since most system indices and data streams are small, so no need to be able to cancel this?

Yes

>This request is cancellable if you give it the parent task id

SystemIndexMigrator uses `ParentTaskAssigningClient`, I believe it should set the parent task id to the requests

masseyke commented:
OK, I think I see two bugs here, one minor and one less so. First, the minor one:
I see this in my logs about half the time I do an upgrade:
```
[2025-03-07T14:00:12,920][ERROR][o.e.c.m.SystemIndexMetadataUpgradeService] [v7.17.28-1] System index metadata upgrade failed java.lang.NullPointerException: Cannot invoke "org.elasticsearch.cluster.metadata.IndexMetadata.getIndex()" because "indexMetadata" is null
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.metadata.SystemIndexMetadataUpgradeService.shouldBeSystem(SystemIndexMetadataUpgradeService.java:171)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.metadata.SystemIndexMetadataUpgradeService$SystemIndexMetadataUpgradeExecutor.updateIndices(SystemIndexMetadataUpgradeService.java:242)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.metadata.SystemIndexMetadataUpgradeService$SystemIndexMetadataUpgradeExecutor.execute(SystemIndexMetadataUpgradeService.java:209)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:1075)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:1038)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService.executeAndPublishBatch(MasterService.java:245)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.lambda$run$2(MasterService.java:1691)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.action.ActionListener.run(ActionListener.java:452)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService$BatchingTaskQueue$Processor.run(MasterService.java:1688)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService$5.lambda$doRun$0(MasterService.java:1283)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.action.ActionListener.run(ActionListener.java:452)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.cluster.service.MasterService$5.doRun(MasterService.java:1262)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1044)
        at org.elasticsearch.server@8.19.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
        at java.base/java.lang.Thread.run(Thread.java:1447)
```
Here's why I think it happens:

- I have a system data stream, `.fleet-actions-results` that was created in v7. I run /_migration/system_features in 8.19.
- The system migration sees that it needs to upgrade the fleet data stream `.fleet-actions-results`.
- The system migration calls the reindex data stream API.
- When the reindex data stream API finishes, it swaps the new index into the `.fleet-actions-results data stream`, and the old one out. This creates a cluster state change
- The system migration sees the cluster state change since SystemIndexMetadataUpgradeService is a ClusterStateListener. It sees that there is a v7 system index named `.ds-.fleet-actions-results-2025.03.07-000001` that is not in a data stream. This is a system index that we need to update!
- The reindex data stream API deletes `.ds-.fleet-actions-results-2025.03.07-000001` now that we no longer need it
- The system migration API is attempting to reindex `.ds-.fleet-actions-results-2025.03.07-000001`. But it has been deleted. So we get the exception above.

This exception is harmless. It's just happening because the index has been deleted in between when we decided we needed to handle it and when we went to handle it. I think we ought to just have [SystemIndexMetadataUpgradeExecutor::updateIndices to continue in the for loop if indexMetadata is null](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/SystemIndexMetadataUpgradeService.java#L243).

masseyke commented:
The other bug I think I'm seeing is that system indices become unreadable after upgrade. If I try to read `.fleet-actions-results` before the upgrade (with a header saying that I am the fleet product), it works fine. After doing an upgrade, making requests to the same index in the same way returns:
```
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"Data stream(s) [.fleet-actions-results] may not be accessed by product [fleet]"}],"type":"illegal_argument_exception","reason":"Data stream(s) [.fleet-actions-results] may not be accessed by product [fleet]"},"status":400}
```
I think the problem is that [system data streams can only have indices starting with `.ds-<data_stream_name>-`](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/indices/SystemDataStreamDescriptor.java#L134.), but the upgrade replaces the original index with one named `.migrated-ds-.fleet-actions-results-2025.03.07-000001`.

JVerwolf commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SystemIndexMigrator.java:
```diff
@@ -575,12 +619,191 @@ private void reindex(SystemIndexMigrationInfo migrationInfo, ActionListener<Bulk
         migrationInfo.createClient(baseClient).execute(ReindexAction.INSTANCE, reindexRequest, listener);
     }
 
+    private void migrateDataStream(
```
IIUC, this doesn't block the datastream from receiving writes while the migration is in progress.  Should it?

alexey-ivanov-es replied:
https://github.com/elastic/elasticsearch/pull/123926#issuecomment-2715421657

JVerwolf reviewed (commented):
This is looking really close to done.  

My only remaining question with respect to design is whether we do/should add a block to the datastream to prevent write loss during a migration?

My other concern is how to have confidence with this huge change-set.  I've reviewed this carefully, but 1600 changes is a lot, even when we are careful. Maybe we can have the help of fleet/datastore to manually test an upgrade before the release is cut?  

Also, you're a coding machine. Nice work!

alexey-ivanov-es commented:
> My only remaining question with respect to design is whether we do/should add a block to the datastream to prevent write loss during a migration?

As I understand - no. The first step in data stream migration is to do a rollover, which creates a new write index for the data stream and all write requests will be send to it. And this new index was created with the current version, so doesn't need migration. @masseyke correct me if I'm wrong

>Maybe we can have the help of fleet/datastore to manually test an upgrade before the release is cut?

Yes, I've already asked them to retest their system indices migration when this PR is merged: https://github.com/elastic/ingest-dev/issues/4826

alexey-ivanov-es commented on x-pack/plugin/migrate/src/main/java/org/elasticsearch/system_indices/task/SingleFeatureMigrationResult.java:
```diff
@@ -82,7 +82,7 @@ public boolean succeeded() {
      */
     @Nullable
-    public String getFailedIndexName() {
+    public String getFailedResourceName() {
```
>maybe in a follow-up PR to do a bunch of renaming

Yes, I don't want to add another 50 files to this PR due to renaming

alexey-ivanov-es committed:
Address review comments

parkertimmins commented:
@alexey-ivanov-es 
>As I understand - no. The first step in data stream migration is to do a rollover, which creates a new write index for the data stream and all write requests will be send to it. And this new index was created with the current version, so doesn't need migration. 

That is correct: if the write index needs to be upgraded, we first do a rollover, so that the new write index has version 8.x .  Most writes should then go to the new write index, but it is still possible that writes go to the other backing indices if the backing indices are written too directly (rather than through the data stream name). Because of this, we actually do set a read-only block on the source backing indices while doing the reindex.

Commit:
[CI] Auto commit changes from spotless

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

alexey-ivanov-es committed:
Fix bugs found during review

alexey-ivanov-es commented:
@masseyke 
>System index metadata upgrade failed java.lang.NullPointerException

I've added protection from that NPE. And I'll also create a separate ticket to review how we handle hidden/system status for data stream backing indices during their lifecycle.

>The other bug I think I'm seeing is that system indices become unreadable after upgrade. If I try to read .fleet-actions-results before the upgrade (with a header saying that I am the fleet product), it works fine. After doing an upgrade, making requests to the same index in the same way returns:

I've added a workaround for this. To properly fix the issue, we have a ticket: ES-10870

alexey-ivanov-es commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java:
```diff
@@ -228,7 +228,7 @@ public class InternalUsers {
                         ILMActions.RETRY.name()
                     )
-                    .allowRestrictedIndices(false)
+                    .allowRestrictedIndices(true)
```
My understanding is that if someone tries to reindex a system data stream directly using the reindexing API without the necessary permissions, it will be blocked before any transport action is triggered. However, I haven't verified this - I'll check this tomorrow

alexey-ivanov-es committed:
Attempt to fix bwc tests running against v8

alexey-ivanov-es committed:
fix build

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

alexey-ivanov-es committed:
Attempt to fix bwc tests

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

masseyke commented:
This looks pretty good to me. A couple of questions:

This ought to go into `main` and `9.0`, right? We don't need it right now, but this would probably be useful to build on when we deal with the 9->10 migration. If nothing else, we'll want the changes from [this commit](https://github.com/elastic/elasticsearch/pull/123926/commits/41b51923208cdf77f4d9d1891af4a224e516b48b) because the system data streams will blow up in 9.x without them.

> Maybe we can have the help of fleet/datastore to manually test an upgrade before the release is cut?

We can just add migration of the fleet data stream to DataStreamsUpgradeIT (or any other upgrade IT), right?

alexey-ivanov-es committed:
Another attempt to fix bwc tests

alexey-ivanov-es commented:
>This ought to go into main and 9.0, right? We don't need it right now, but this would probably be useful to build on when we deal with the 9->10 migration. If nothing else, we'll want the changes from [this commit](https://github.com/elastic/elasticsearch/pull/123926/commits/41b51923208cdf77f4d9d1891af4a224e516b48b) because the system data streams will blow up in 9.x without them.

Yes, this will be in 9.x/main branches with small modifications.

>We can just add migration of the fleet data stream to DataStreamsUpgradeIT (or any other upgrade IT), right?

I've already created [SystemIndicesUpgradeIT](https://github.com/elastic/elasticsearch/blob/1aba6780e1012ddf7f717bce9ec4089b5431477a/x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SystemIndicesUpgradeIT.java), but I think it would be useful to retest it manually as well

masseyke commented:
> I've already created [SystemIndicesUpgradeIT](https://github.com/elastic/elasticsearch/blob/1aba6780e1012ddf7f717bce9ec4089b5431477a/x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SystemIndicesUpgradeIT.java), but I think it would be useful to retest it manually as well

Ah, sorry, I had missed the addition of that test!

JVerwolf reviewed (approved):
LGTM, provided that there's been a bit of regression testing with security enabled to ensure we aren't breaking regular system index migration.

masseyke commented:
> LGTM, provided that there's been a bit of regression testing with security enabled to ensure we aren't breaking regular system index migration.

It looks like the new SystemIndicesUpgradeIT covers both a system index and a system data stream, and it runs with security enabled.

masseyke commented:
SystemIndicesUpgradeIT just failed for me with the following seed (mentioning this here b/c CI might pick a seed that succeeds for the PR build):
```
./gradlew ':x-pack:qa:rolling-upgrade:v7.17.28#upgradedClusterTest' -Dtests.class="org.elasticsearch.upgrades.SystemIndicesUpgradeIT" -Dtests.seed=68EB9E4AB8CF1573 -Dtests.bwc=true -Dtests.locale=et -Dtests.timezone=Europe/Tiraspol
```

masseyke commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SystemIndicesUpgradeIT.java:
```diff
@@ -0,0 +1,194 @@
+        Request request = new Request("PUT", "/_security/role/" + name);
+        request.setJsonEntity(
+            "{ \"cluster\": [\"cluster:admin/migration/post_system_feature\", \"cluster:admin/migration/get_system_feature\"],"
+                + " \"indices\": [ { \"names\" : [ \""
+                + indexOrDataStream
+                + "\"], \"privileges\": [ \"manage\" ] } ] }"
```
This part, and the passed-in indexOrDataStream could be removed, since the only permissions needed for this to work are the cluster-level permissions.

alexey-ivanov-es commented:
@masseyke 
>SystemIndicesUpgradeIT just failed for me

Do you have an exception it failed with? It might be failing with ~ "status is IN_PROGRESS, but expected NO_MIGRATION_NEEDED" because migration sometimes takes more than 10 seconds, so I am going to increase timeout in `assertBusy`

masseyke commented:
> @masseyke
> 
> > SystemIndicesUpgradeIT just failed for me
> 
> Do you have an exception it failed with? It might be failing with ~ "status is IN_PROGRESS, but expected NO_MIGRATION_NEEDED" because migration sometimes takes more than 10 seconds, so I am going to increase timeout in `assertBusy`

I thought I deleted my comment quickly enough, but apparently I didn't. :) It was failing for me because I didn't have your most recent commits. Once I pulled those, the problem went away.

alexey-ivanov-es committed:
Remove unnecessary permissions and increase timeout

alexey-ivanov-es committed:
Merge branch '8.x' into ES-10934

alexey-ivanov-es committed:
Mute tests
